JLI Support

jm4rtin opened the pull request:
This change adds JLI support to GCC. This will accept the jli_call_fixed and jli_call_always pragmas used by Metaware. This implementation should be binary compatible with Metaware (if using the default option -mjli-func-addr=compat).

There are three approaches to obtaining the address of a function inside the JLI table:
1. -mjli-func-addr=compat will use a Metaware compatible mode. When using jli_call_fixed this will use the address of the function originally placed into the table at compile time. When using jli_call_always this will use the address of the JLI entry. The JLI table may not be moved when using jli_call_always with this method.
2. -mjli-func-addr=init will read the jli_base AUX register and offset it to determine the address of the JLI entry. This address will be stored in the desired variable. If the variable is constant or otherwise ends up in the .rodata section, the address of the .jlitab section will be used in place of the jli_base AUX register. This will emit an ARC_32_JLI relocation entry that may be used to perform the proper relocation utilizing the jli_base AUX register during loading. This method requires the call be made in kernel mode (because of the AUX register read).
3. -mjli-func-addr=always will create a stub function that reads the jli_base AUX register and offset to determine the address of the JLI entry. The address of the stub function will be stored in the desired variable so the JLI table may be moved after the variable has been set with the function address. This method requires the call be made in kernel mode (because of the AUX register read).

The following relocation types were added to GCC/binutils:
1. ARC_JLI_SECTOFF is a relocation type in Metaware that is now used by GCC as well to adjust the index of function calls to functions defined with #pragma jli_call_always during linking.
2. ARC_JLI_SECTOFF_SIMM12 is used to add the appropriate offset to the address when using -mjli-func-addr=init to get an address for a function in the JLI table during code execution.
3. ARC_JLI_32 is used to add the appropriate offset to the address when using -mjli-func-addr=init to get an address for a function in the JLI table during program load.

This implementation of the JLI feature will allow the use of both jli_call_fixed and jli_call_always as long as all jli_call_fixed functions are using the continuous block starting at index 0 and are linked before all other JLI entries. This can be done with the default linker script if the object file that contains the jli_call_fixed entries matches the object filename pattern "jlitab*.o".

claziss commented:
Thank you for ur contribution. I will give some first level review using the github system. Please let us know if you would like to take in our feedback or not.

claziss commented on bfd/elflink.c:
```diff
@@ -13564,6 +13564,13 @@ _bfd_elf_section_already_linked (bfd *abfd,
 
   flags = sec->flags;
 
+  /* Check if the section type is SHT_COMDAT and the SEC_LINK_ONCE flag is not
```
This is mainline code, not sure if you should add this piece of code here. All the backend code should be in arc specific bfd file.

compomega replied:
The section(s) are included for any file that uses JLI calls (the entry + stub). Only one copy of these sections (with the same name) should be included and the other discarded. The only way I see for an ELF file that has this behavior is the .gnu.linkonce.* sections; however, using this would change the names used as it may not be compatible with Metaware. If you have another suggestion on how to do this, please let me know.

jm4rtin replied:
The section(s) are included for any file that uses JLI calls (the entry + stub). Only one copy of these sections (with the same name) should be included and the other discarded. The only way I see for an ELF file that has this behavior is the .gnu.linkonce.* sections; however, using this would change the names used as it may not be compatible with Metaware. If you have another suggestion on how to do this, please let me know.

claziss replied:
That may be solved using groups and comdat. From as manual:

.section name , "flags"G, @type, GroupName[, linkage]
The GroupName field specifies the name of the section group to which this particular
section belongs. The optional linkage field can contain:
- **comdat** indicates that only one copy of this section should be retained
- **.gnu.linkonce** an alias for comdat

Anyhow, we need to relay on gnu.linkonce as this is the default gnu way of handling this. A 100% metware compatibility may not be possible. Alternatively, we can ask metaware guys to support gnu like names.

claziss commented on gas/config/obj-elf.c:
```diff
@@ -704,7 +704,7 @@ obj_elf_change_section (const char *name,
 #endif
 
-  if (linkonce)
+  if (type == SHT_COMDAT || linkonce)
```
I don't think it is a good place for this code, as you add backend specific code into general part.

compomega replied:
See my first comment.

jm4rtin replied:
See first comment.

claziss commented on gas/testsuite/gas/arc/taux.d:
```diff
@@ -16,6 +16,7 @@
 .* U d2l
 .* U dpfp_status
 .* U fp_status
+.* U jli_base
```
jli may not be present for arc700

compomega replied:
The test was failing after the jli_base AUX register was corrected for only EMv2. Does this not mark it as undefined for the arc700 so the test will pass without that AUX register showing up?

jm4rtin replied:
The test was failing after the jli_base AUX register was corrected for ARCv2. Does this not mark it as undefined for the arc700 so the test will pass without the AUX register showing up?

claziss replied:
We can make a regex to handle this. We will fix it later. BTW, we need so jli tests as well. Please can u prepare/make some of them.

claziss commented on include/elf/common.h:
```diff
@@ -464,6 +464,7 @@
 #define SHT_REL		9		/* Relocation entries, no addends */
 #define SHT_SHLIB	10		/* Reserved, unspecified semantics */
 #define SHT_DYNSYM	11		/* Dynamic linking symbol table */
+#define SHT_COMDAT   12    /* Common data */
```
This is something which affects all backends, needs to be moved.

compomega replied:
See my first comment.

jm4rtin replied:
See first comment.

claziss commented on ld/scripttempl/arclinux.sc:
```diff
@@ -483,6 +483,13 @@ cat <<EOF
   ${TEXT_PLT+${PLT_NEXT_DATA-${PLT}}}
   ${TINY_READONLY_SECTION}
+
+  .jlitab       ${RELOCATING-0} :
```
jli should not be used for linux.

jm4rtin replied:
It can be removed.

claziss commented on opcodes/arc-regs.h:
```diff
@@ -243,6 +242,7 @@ DEF (0x236, ARC_OPCODE_ARCALL, NONE, ap_amm7)
 DEF (0x237, ARC_OPCODE_ARCALL, NONE, ap_ac7)
 DEF (0x278, ARC_OPCODE_ARCALL, NONE, pct_control)
 DEF (0x279, ARC_OPCODE_ARCALL, NONE, pct_bank)
+DEF (0x290, ARC_OPCODE_ARCv2EM, NONE, jli_base)
```
I would add also jli for ARCHS

jm4rtin replied:
Yes I can add that.

claziss commented on opcodes/arc-tbl.h:
```diff
@@ -571,6 +571,9 @@
 { "addsdw", 0x2EE87F80, 0xFFFF7FE0, ARC_OPCODE_ARC600 | ARC_OPCODE_ARC700, ARITH, NONE, { ZA, LIMM, LIMMdup }, { C_F, C_CC }},
 
+/* add2_jlioff<.f> b,b,s12 00100bbb10010101FBBBssssssSSSSSS.  */
+{ "add2_jlioff", 0x20950000, 0xF8FF0000, ARC_OPCODE_ARC600 | ARC_OPCODE_ARC700 | ARC_OPCODE_ARCv2EM | ARC_OPCODE_ARCv2HS, ARITH, NONE, { RB, RBdup, SIMM12_20_JLIOFF }, { C_F }},
```
Are you absolutely sure this instruction is valid for arc6xx and A7?

jm4rtin replied:
This is a copy of the add2 instruction that takes a symbol and emits a relocation entry. As such, the instruction should support those; however, this form of it could be limited to just ARC_OPCODE_ARCv2EM | ARC_OPCODE_ARCv2HS.

claziss replied:
Then, this mod is not required. We will use the normal add2 instruction plus a relocation. E.g.: add2 rx,ry,@sym@jli. IMHO, it is cleaner in this way.

claziss commented on opcodes/arc-tbl.h:
```diff
@@ -8043,7 +8046,10 @@
 
 /* jli_s u10 010110uuuuuuuuuu.  */
-{ "jli_s", 0x00005800, 0x0000FC00, ARC_OPCODE_ARCv2EM | ARC_OPCODE_ARCv2HS, JUMP, CD1, { UIMM10_6_S }, { 0 }},
+{ "jli_s", 0x00005800, 0x0000FC00, ARC_OPCODE_ARCv2EM | ARC_OPCODE_ARCv2HS, JUMP, ARC_CD, { UIMM10_6_S }, { 0 }},
```
Why you use ARC_CD for the subclass. Do u really need both jli_s variants?

jm4rtin replied:
For some reason when I was compiling the code after merging it into arc-2016.09 it would no longer accept the opcode with CD1 despite compiling with -mcpu=em4 -mcode-density. This was working for arc-2016.03 so I am not sure why it stopped working.

claziss replied:
Hmm, probably because we may have an issue with the gcc driver. But this should not be the reason for this mod. I've checked the jli_s instruction with the assembler only, and it works correctly. And, by all means, ARC_CD is wrong, when you are not sure about the instruction SUBCLASS, we should use NONE.

jm4rtin commented on ld/scripttempl/elfarc.sc:
```diff
@@ -262,6 +262,11 @@ cat <<EOF
   } =${NOP-0}
 
   ${DATA_PLT-${BSS_PLT-${PLT}}}
+  .jlitab       ${RELOCATING-0} :
```
Should this be removed as well as the Linux linker changes since JLI is only supported by ARCv2 processors?

claziss replied:
No, you can let it here, as we use elfarc.sc by default.

jm4rtin commented on gas/config/obj-elf.c:
```diff
@@ -1135,7 +1137,8 @@ obj_elf_section (int push)
 		      linkonce = 1;
 		    }
 		}
-	      else if (strncmp (name, ".gnu.linkonce", 13) == 0)
```
I did not mean to leave this change in and will remove it.

claziss commented:
I've asked my colleague to give some input, probably we will talk Monday about this issue. This patch is worth to go upstream but we need to make sure those guys will not reject it. Good work otherwise.

claziss commented on bfd/bfd-in2.h:
```diff
@@ -3737,6 +3737,8 @@ pc-relative or some form of GOT-indirect relocation.  */
   BFD_RELOC_ARC_S21H_PCREL_PLT,
   BFD_RELOC_ARC_NPS_CMEM16,
   BFD_RELOC_ARC_JLI_SECTOFF,
+  BFD_RELOC_ARC_JLI_SECTOFF_SIMM12,
```
Are those additional relocations? Why do we need them, I am not aware MWDT has them.

jm4rtin replied:
As described in the original description of the pull request, these additional relocation entries are used for the additional methods of obtaining the address of a function that is in the JLI table (that do not exist in Metaware right now). These utilize the jli_base AUX register and offset it.

ARC_JLI_32 is for initial values of a global or static variable. Take the following example:
`#pragma jli_call_always(TestFunction)`
`extern void TestFunction(void);`
`static void *gSomeFunction = &TestFunction;`

In this case, during program load, gSomeFunction normally gets set to the address of TestFunction. Instead, with -mjli-func-addr=init this will link to the address of the JLI entry inside the JLI table. Alternatively, the relocation entry could be used during program load to re-write the initial contents as the jli_base AUX register + the section offset for the entry (which is the more preferred formula). This would take into account a move of the JLI table between initial boot and the loading of this program.

ARC_JLI_SECTOFF_SIMM12 is used for the assignment of a function address for a function in the JLI table. Take the following example:
`#pragma jli_call_always(TestFunction)`
`extern void TestFunction(void);`
`void *gSomeFunction;`
`/* some code here */`
`gSomeFunction = &TestFunction;`

At this point, normally gSomeFunction gets the address of TestFunction; however, with -mjli-func-addr=init it will get the address of the entry in the JLI table where the JLI table address is determined by reading the jli_base AUX register (in case it moved). In this case, an lr is followed by a relocated add2 whose immediate value depends on where in the JLI table the enty was linked.

claziss commented on include/opcode/arc-func.h:
```diff
@@ -272,9 +272,40 @@ ATTRIBUTE_UNUSED static unsigned
+  extern bfd_signed_vma jli_base;
+
+  insn = insn & ~0xffffffff;
+  insn |= (((int)(jli_base + value) >> 0) & 0xffffffff) << 0;
```
This should be in the relocation formula. This file is only for insert functions used by relocs, and it should not attempt changing the parameter value.

claziss commented on opcodes/arc-tbl.h:
```diff
@@ -8043,10 +8046,10 @@
 
 /* jli_s u10 010110uuuuuuuuuu.  */
-{ "jli_s", 0x00005800, 0x0000FC00, ARC_OPCODE_ARCv2EM | ARC_OPCODE_ARCv2HS, JUMP, CD1, { SIMM25_6_S }, { 0 }},
+{ "jli_s", 0x00005800, 0x0000FC00, ARC_OPCODE_ARCv2EM | ARC_OPCODE_ARCv2HS, JUMP, ARC_CD, { UIMM10_6_S_JLIOFF }, { 0 }},
```
Probably we need to keep only one variant, and that may be this one, after we correctly use CD1

claziss commented on include/elf/arc-reloc.def:
```diff
@@ -497,3 +497,24 @@ ARC_RELOC_HOWTO(ARC_NPS_CMEM16, 78, \
+                32, \
+                replace_word32_jli, \
+                bitfield, \
+                ( ( S - JLI ) ))
```
As far as I can see this formula combined with the op in replace_word32_jli says that ARC_JLI_32 is actually (S - JLI) + JLI = S. Is it correct?

jm4rtin replied:
In this case, yes (which I have changed to simply S and using replace_word32); however, if the program is being loaded and relocated the loading code may wish to use JLI_BASE + ( S - JLI ) which could differ from the section's address. Here, JLI_BASE would be read from the AUX register. For this reason I wanted a different relocation type.

claziss commented:
Here is an high-level overview of what are my comments to your patch:

1. Relocation ARC_JLI_32/ARC_SECTOFF_SIMM12 needs to be cleanup, i.e., computation should happen in the reloc formula;

2. The mainline mods are unnecessary, we need to use the gnu comdat or gnu.linkonce features. If this brings issues with metware produced binaries, we need to understand how we can address them without those changes;

3. add2_jlioff is not required. We need to add @jli reloc in assembler.

4. I miss assembler/linker tests, we need to create them for all three relocations introduced. Those tests are simple asm files and their expected object output.

Maybe it will be wise if we can have a chat about those changes somewhere soon. Once the binutils issues are cleaned out, we move to gcc patch.

claziss commented on ld/scripttempl/elfarc.sc:
```diff
@@ -264,6 +264,7 @@ cat <<EOF
   ${DATA_PLT-${BSS_PLT-${PLT}}}
   .jlitab       ${RELOCATING-0} :
   {
+    jlitab.o(.jlitab*)
```
Who is producing this file, where is the source of this file.

jm4rtin replied:
That got changed to jlitab*.o so it doesn't kill the linker if it doesn't exist. This file would be made by the user with jli_call_fixed and is in the default linker script to ensure fixed entries come before always entries in the case the two types are mixed.

jm4rtin commented:
To your main comments:

1. I cleaned this up in my working copy which I will push to Github soon . 
2. I've tried out the other .section declaration (which works for my need) and removed these changes.
3. Is there a way to recognize the @jli prefix to a symbol name and emit a certain relocation entry? If so, I'll look into adding that.
4. I'll look at that. I haven't had the best of luck getting the GCC tests to play nice with NSIM but the binutil tests are running for me.

jm4rtin commented:
I figured out how to add @jli but now I have a question. Doesn't the instruction still need a version that accepts a symbol name (which I can make emit a relocation entry anyway)? Or is there some way to get the original instruction to accept a symbol name?

claziss commented:
In gcc, u can just add @jli when emitting the add2 instruction which is part of the bultin. If you want to be more advanced u can check how @pcl is done. By all means, we will make it proper when time arrives. U can just use the quick one and add @jli directly when emitting the instruction.

claziss commented:
Oh, I think I miss understood u, you talk about the assembler. No the instruction shouldn't need any special handling. Again, follow the @pcl or @sda implementations.

jm4rtin commented:
I'm still not seeing how to do it. I'm trying it out with the jli_s call right now. I have GCC emit the following assembly:
`jli_s @__jli.TestFunction2@jli`

I then remove the other jli_s assembly function I added that accepts a symbol:

`/* jli_s u10 010110uuuuuuuuuu.  */`
`{ "jli_s", 0x00005800, 0x0000FC00, ARC_OPCODE_ARCv2EM | ARC_OPCODE_ARCv2HS, JUMP, CD1, { UIMM10_6_S_JLIOFF }, { 0 }},`

So that only this one remains:

`/* jli_s u10 010110uuuuuuuuuu.  */`
`{ "jli_s", 0x00005800, 0x0000FC00, ARC_OPCODE_ARCv2EM | ARC_OPCODE_ARCv2HS, JUMP, CD1, { UIMM10_6_S }, { 0 }},`

The assembler will not match this version of the instruction and complain:
`inappropriate arguments for opcode 'jli_s'`

Is the second version needed or is there a way to get this to match the first one?

jm4rtin commented:
I could also emit this assembly instead and use no additional instructions in the assembler:
`.reloc ., R_ARC_JLI_SECTOFF, __jli.TestFunction2`
`jli_s 0`

I'm just curious how to get the @jli method to work.

claziss commented:
@JLI is intended only for the add2. For the jli_s instruction, I would keep ur implementation. Sorry, maybe I was not clear enough.

claziss commented:
if u want to use @jli for both jli_s and add2 instruction may be a bit more complicated and it is not needed , I would say.

jm4rtin commented:
Then is there something set in the add2 instruction that allows it to accept a symbol name or an @identifier@relocation_type operand?

claziss commented:
IMHO, add2 r0,pcl,@sym@pcl should work

jm4rtin commented:
Perhaps it does and I will try it; however, I was curious what part of the code makes it work?

claziss commented:
gas/config/tc-arc.c. For S12 we need to add a reloc in the SIMM12_20 definition. Try an inernal one:
{12, 0, -SIMM12_20, ARC_OPERAND_SIGNED, insert_simm12_20, extract_simm12_20},

claziss commented:
so,

1. define a machine specific operand: #deifine O_jli O_md12
2. Update O_last
3. add it in arc_reloc_op, as non complex expression
4. update debug_exp
5. update find_opcode_match
6. update assemble_insn

That should be it

jm4rtin commented:
It looks like when I do this it emits a different variant of add2 that has a 32-bit immediate. It seems like I'm back to the same problem that @identifier@jli isn't matching the instruction I want.

Also, this comment was not very clear to me:
gas/config/tc-arc.c. For S12 we need to add a reloc in the SIMM12_20 definition. Try an inernal one:
{12, 0, -SIMM12_20, ARC_OPERAND_SIGNED, insert_simm12_20, extract_simm12_20},

claziss commented:
Right sorry, too fast. So, to add @jli u need to change the assembler the tc-arc.c file as sketch above.

To be able have a relocation for s12, we need to add that s12 can accept a relocation, and we do this in arc-opc.c, the SIMM12_20 definition as indicated above.

claziss commented:
If u still have difficulties, plz make a branch and indicate it to me. I will try to help u on the live code

claziss commented on gas/config/tc-arc.c:
```diff
@@ -1887,6 +1890,7 @@ find_opcode_match (const struct arc_opcode_hash_entry *entry,
 		    case O_gotoff:
 		    case O_gotpc:
 		    case O_pcl:
+		    case O_jli:
```
Here we need to move O_jli in a stand alone case, as all other relocs are require for LIMM. So:
case O_jli:
if (!(operand->flags & ARC_OPERAND_SIGNED))
  goto match_failed;
break;

claziss replied:
Maybe you would also like to check for the operand number of bits, just to be absolutely sure u match the S12 immediate.

claziss commented on gas/config/tc-arc.c:
```diff
@@ -3820,6 +3824,11 @@ assemble_insn (const struct arc_opcode *opcode,
 				  operand->default_reloc);
 	      break;
+	    case O_jli:
+	      reloc = find_reloc ("jli", opcode->name,
```
You need to add this:
reloc = ARC_RELOC_TABLE (t->X_md)->reloc;
	      break;

jm4rtin replied:
What does find_reloc do different then (I was just borrowing the O_sda case)?

claziss replied:
find_reloc is more complex, for example you may use @jli for various instructions, you need to generate for each instruction a specific relocation. This selection which relocation for which instruction can be done/added by find_reloc. In ur case we have a single reloc as indicated in the ARC_RELOC_TABLE.

jm4rtin commented:
I see, so relocation types that are negative are handled by this special list and need to be resolved into a proper relocation type with a positive ID? I'm getting an assertion that the relocation type is still internal:

Assertion failure in apply_fixups at gas/config/tc-arc.c:1338

Here is my sandbox I'm testing some code with: https://github.com/jm4rtin/sandbox
Here is the binutils-gdb branch: https://github.com/emmicro-us/binutils-gdb/tree/arc-2016.09-em
Here is the gcc branch: https://github.com/emmicro-us/gcc/tree/arc-4.8-dev-em

Also, tc-arc.c has an interesting mix of tabs and spaces that was making for odd diffs when my text editor was converting tabs into spaces for new lines.

jm4rtin commented:
Looks like the pull request has been updated with my changes :)

claziss commented on opcodes/arc-opc.c:
```diff
@@ -1695,7 +1695,7 @@ const struct arc_operand arc_operands[] =
   /* SIMM12_20 mask = 00000000000000000000111111222222.  */
 #define SIMM12_20	(UIMM6_20 + 1)
-  {12, 0, 0, ARC_OPERAND_SIGNED, insert_simm12_20, extract_simm12_20},
+  {12, 0, -SIMM12_20, ARC_OPERAND_SIGNED, insert_simm12_20, extract_simm12_20},
```
This is OK, so what you say here is that S12 can accept a relocation.-SIMM12_20 will instruct the assembler ARC backend that the default relocation is an internal dummy reloc having insert_simm12_20 as the relocation solver.

jm4rtin replied:
There is a problem with this. When using -SIMM12_20 on an add2 that does not have @identifier@jli it throws an assertion because the internal relocation was not handled. As such, newlib won't even compile. Is there something extra I must do to handle this relocation before the assertion happens? I could instead check for -SIMM12_2 (which isn't defined in tc-arc.c) as the relocation type and replace it with BFD_RELOC_ARC_NONE (I would still need to define SIMM12_20 first - I used a hard coded number to test) :

```
diff --git a/gas/config/tc-arc.c b/gas/config/tc-arc.c
index c71bce4..c24e262 100644
--- a/gas/config/tc-arc.c
+++ b/gas/config/tc-arc.c
@@ -3879,6 +3879,11 @@ assemble_insn (const struct arc_opcode *opcode,
          if (insn->nfixups >= MAX_INSN_FIXUPS)
            as_fatal (_("too many fixups"));

+         if (reloc == -SIMM12_20)
+         {
+           reloc = BFD_RELOC_ARC_NONE;
+         }
+
          struct arc_fixup *fixup;
          fixup = &insn->fixups[insn->nfixups++];
          fixup->exp = *t;
```

Is there a better way?

claziss replied:
We hit an backend limitation, which I will need to address it latter on. Namely, the selection of the right opcode when the mnemonic demands a relocation. In our case we have a reloc for s12 and a reloc for limm. The assembler picks the first one i.e. s12 reloc. Hence, the issue.
Anyhow, I did fixed for u, and as far as I can see there shouldn't be any error. You can find the patch here:
https://github.com/foss-for-synopsys-dwc-arc-processors/binutils-gdb/tree/jli

claziss commented:
Yeah, github is really smart. This is one of the reasons why I find it nice to work with it.

jm4rtin commented:
It seems to work now. Thanks for the help. Aside from adding a test into the suite, how does it look?

claziss commented:
My pleasure. For the test check the gas ones here: gas/testsuite/gas/arc/... Or any other folder. 
For the linker (here all relocs are solved, hence you can do more trics to crash the linker), you can check ld/testsuite/ld-arc.
Build everything and run make check-gas or make check-ld to see the results.

claziss commented on include/elf/arc-reloc.def:
```diff
@@ -497,3 +497,24 @@ ARC_RELOC_HOWTO(ARC_NPS_CMEM16, 78, \
+                bitfield, \
+                ( ( ( S - JLI ) >> 2 ) ))
+
+ARC_RELOC_HOWTO(ARC_JLI_32, 80, \
```
I am not really convinced that we did a good job here. Please can u explain once again why do u need this reloc. I find a bit strange a reloc can require two formulas to be solved.

jm4rtin replied:
ARC_JLI_SECTOFF_SIMM12 was for the add2 instruction. Here, the address of the function in the JLI table is taken as the offset into the JLI section ((S - JLI) >> 2) (the shift is because of add2) plus the base of the JLI section _when this code is executed_. If you don't move the JLI table it equates to S. If you do move the JLI table, it equates to (JLI_BASE + (S - JLI)) where JLI_BASE is the contents of the AUX register and not the original location in the ELF file.

ARC_JLI_32 is a bit more tricky. When you initialize a static/global variable with the address of a function in the JLI table, the linker must then solve for this address. In this case, the linker can't solve (JLI_BASE + (S - JLI)) but it can solve (JLI + (S - JLI)) or S. That means it's up to the program loader to make sure this is properly changed if the JLI table was moved. For a main application you may want to simply set the jli_base register to "JLI" but if you are loading in multiple targets, you may not want to alter the jli_base AUX register and simply relocate all function pointers to ensure they point into the pre-existing JLI table.

claziss replied:
Thanks, JLI_SECTOFF_SIMM12 is clear to me.
However, it seems to me, this ARC_JLI_32 is also a sort of dynamic reloc. Also, I see you assume JLI_BASE doesn't change its value at run time, as you need it to be constant to solve the ARC_JLI_32 at load time. But then, who stops you to define the value of JLI_BASE when linking and obtaining directly the value which you will anyhow obtain during solving phase of the loader.
Another issue is we need to nicely generate this reloc, rather than spelling it out.

claziss commented:
Hi,
I have made for u a number of modifications in the https://github.com/foss-for-synopsys-dwc-arc-processors/binutils-gdb/tree/jli branch.
One of them handles the emitting of the ARC_JLI_32 reloc. Hence, you can emit from gcc something like:
.word .@foo@jli.

Please check them, and let me know if they are working.

Finally, you need to add the following missing items:

1.  Add linker/assembler tests;

2.  Add changelogs lines in the commit message.

3. Squash everything in 1-2 commits.

That is it. Please let me know if you have difficulties.

jm4rtin commented:
Alright, I'll let you know. Also working on something internal right now so it may take a little bit of time to finish this off.

claziss commented:
I've merged/cleaned all this pull request into a single commit in the main branch. Once all the internal checks, updates are done, I will close this pull-request.

claziss commented:
New jli support available in 2017.03. Closing
